### DISCUSSION TITLE
Optimistic UI with multiple submits from the same fetcher

### DISCUSSION THREAD

**Discussion by guillaumebrunerie** [remix-run/remix#5026] [Q&A]
I'm trying to implement optimistic UI in a todo-list app, but I can’t figure out how to do it in a way that properly deals with multiple submits from the same `fetcher`.

Let's say we display a list of items and there is an `Add` button to add a new item to the list. It is doable to add optimistic UI to this button, by using the `fetcher.submission` to display a fake item until the "real" one appears, but what if the user has a flaky connection and manages to add more items before receiving an answer from the first one? The problem is that a `fetcher` only tracks the last submission, so if `fetcher.submit` has been called several times you can’t access the older ones, and you can’t know whether they are completed or are still in progress.

I tried to use the native `fetch` instead, until I realized that the promise resolves at the wrong time (it resolves before the page is reloaded by Remix, so we get an unpleasant UI blink).

I've watched the video https://www.youtube.com/watch?v=EdB_nj01C80, that shows the same kind of app, but optimistic UI for adding items was conveniently omitted…

I have in the client the last state sent by the server (the loader data), so if I could just know which actions (= calls to `fetcher.submit`) are still pending (= older than the current loader data), it would be perfect.

**Comment by kiliman** [comment]
Fetchers are meant for single use. You should create a separate component for adding new items. That component can have its own fetcher.

**Reply by juniorforlife** [reply]
@kiliman in my case, I have a list of items, each item has a button to send a POST request to the server. This request will affect the UI of all the other items in the list. 
I can do optimistic UI for the current card by using useFetcher to receive the data. 
However, the other items cannot see fetcher.data because they are not the ones that trigger the submission. Can you guide me how to achieve this?

**Reply by sergiodxa** [reply]
You can use useFetchers (in plural) to get the list of active fetchers. Then find the fetcher in progress of sending a request and show something on the UI.

**Reply by juniorforlife** [reply]
@sergiodxa I tried useFetchers but when I console.log fetchers it always returns an empty array. How do I use fetcher.submit like useFetcher?

**Reply by juniorforlife** [reply]
I made a video to better explain my problem
https://www.youtube.com/watch?v=GRM9Ez87mNo
